src/core/stdc/stdarg.d

```diff
@@ -34,466 +80,297 @@ unittest
-                        ap.offset_fpregs += 16;
-                    }
-                    else
-                    {
-                        p = ap.stack_args;
-                        ap.stack_args += tsize.alignUp;
-                        stack = true;
-                    }
-                }
-                else
-                {   // Passed in regular register
-                    if (ap.offset_regs < 6 * 8 && !stack)
-                    {
-                        p = ap.reg_args + ap.offset_regs;
-                        ap.offset_regs += 8;
-                    }
-                    else
-                    {
-                        p = ap.stack_args;
-                        ap.stack_args += 8;
-                        stack = true;
-                    }
-                }
-                parmn[0..tsize] = p[0..tsize];
-
-                if (arg2)
-                {
-                    if (inXMMregister(arg2))
-                    {   // Passed in XMM register
-                        if (ap.offset_fpregs < (6 * 8 + 16 * 8) && !stack)
-                        {
-                            p = ap.reg_args + ap.offset_fpregs;
-                            ap.offset_fpregs += 16;
-                        }
-                        else
-                        {
-                            if (!stack)
-                            {   // arg1 is really on the stack, so rewind and redo
-                                ap.offset_fpregs = offset_fpregs_save;
-                                ap.offset_regs = offset_regs_save;
-                                stack = true;
-                                goto L1;
-                            }
-                            p = ap.stack_args;
-                            ap.stack_args += arg2.tsize.alignUp;
-                        }
-                    }
-                    else
-                    {   // Passed in regular register
-                        if (ap.offset_regs < 6 * 8 && !stack)
-                        {
-                            p = ap.reg_args + ap.offset_regs;
-                            ap.offset_regs += 8;
-                        }
-                        else
-                        {
-                            if (!stack)
-                            {   // arg1 is really on the stack, so rewind and redo
-                                ap.offset_fpregs = offset_fpregs_save;
-                                ap.offset_regs = offset_regs_save;
-                                stack = true;
-                                goto L1;
-                            }
-                            p = ap.stack_args;
-                            ap.stack_args += 8;
-                        }
-                    }
-                    auto sz = ti.tsize - 8;
-                    (parmn + 8)[0..sz] = p[0..sz];
-                }
+                const arg = va_arg!T(_argptr);
+                //FIXME: assert(arg.value == i);
```

Comment:
This remaining issue here is IMO due to a DMD codegen bug when emitting the varargs. It doesn't work for LDC either; to be honest, I didn't expect non-PODs to be allowed as varargs (only those with dtors are currently disallowed).

Reply:
And now postblits and copyctors are disallowed too.

Nested structs are also passed by invisible ref in gdc, but that doesn't affect va_arg, because as soon as the type symbol escapes, it's no longer "enclosed".